src/efs_parser/BigBed.py

```diff
@@ -105,3 +105,18 @@ def parseLeafDataNode(self, chrmId, start, end, zoomlvl, rStartChromIx, rStartBa
                             x += 1
                     x += 2
         return result
+    def groupRows(self,res):
+        result = pd.DataFrame(columns=res.columns)
+        first_row = True
+        index = 0
+        for _, row in res.iterrows():
+            if first_row:
+                result.loc[index] = row
+                first_row = False
+            else:
+                if row['name'] == result['name'][index]:
+                    result['end'][index] = row['end']
+                else:
+                    index += 1
+                    result.loc[index] = row
+        return result
```

Comment:
within each of the groups, the start is the min of the starts and the end if the max of all the ends. This is so that we are now representing these smaller intervals as one single long interval.

Reply:
let me know if this is confusing

Reply:
This is how the algorithm works:
1 - Fill the first record of the result with the first record of the dataframe
2 - as long as the the name is not changing. update the end of the current result record to make the end equals the end of the current record
3 - Then, when the name changes, a new record is created in the result and it will equal the current record of the dataframe   
4 - Step 2 is repeated and so on

Reply:
we want it make it more efficient by using pandas group by. Iterating rows in pandas is a costly operation. 

wrt intervals 
```
starts - ends
2 - 3 Quies ...
1 - 3 Quies ...
3 - 4 Quies ...
```
in the same group and we need to represent this group as a single interval, we take the min of all the starts (i.e 1) and the max of all the ends (4) and represent that as the new interval. we do min and max incase the rows are not sorted.

```
1-4 Quies ...
```

Reply:
This is a good article on why iterrows is not efficient (also google will get you a lot more results) - https://stackoverflow.com/questions/24870953/does-pandas-iterrows-have-performance-issues#:~:text=Iterating%20through%20pandas%20objects%20is,needed%20%5B...%5D.&text=iterrows%20is%20by%20far%20the,aren't%20much%20better%20either.

but in summary vectorized operations are preferred whenever possible when working with pandas

Reply:
Just to make sure that I understand the requirement correctly.

From Actual data, I found that the start of record n+1 = end of record n
consider the example from file:
https://obj.umiacs.umd.edu/bigwig-files/ENCFF330GHF.bigBed
chr1	10013600	10013800	QuiesG
chr1	10013800	10014000	QuiesG
chr1	10014000	10014200	EnhPois2
chr1	10014200	10014400	QuiesG
chr1	10014400	10014600	QuiesG
chr1	10014600	10014800	QuiesG
chr1	10014800	10015000	EnhPois2
chr1	10015000	10015200	QuiesG
chr1	10015200	10015400	QuiesG

The result as I understand should be:
chr1	10013600	10014000	QuiesG
chr1	10014000	10014200	EnhPois2
chr1	10014200	10014800	QuiesG
chr1	10014800	10015000	EnhPois2
chr1	10015000	10015400	QuiesG

Did I understand correctly?

Reply:
yes thats correct

Reply:
if we use the group by and pick the min and the max for the name the result may look like this:
chr1 10013600 10015400 QuiesG
chr1 10014000 10015000 EnhPois2

Reply:
you can pass in a parameter `sort = False` to groupby so it only groups consecutive rows

Reply:
as an example - 


```
>>> filtered
      chr     start       end       name
375  chr1  10007800  10008000      Quies
376  chr1  10008000  10008200      Quies
377  chr1  10008200  10008400      Quies
378  chr1  10008400  10008600      Quies
379  chr1  10008600  10008800   EnhPois1
380  chr1  10008800  10009000  TssAFlnk2
381  chr1  10009000  10009200  TssAFlnk2
382  chr1  10009200  10009400  TssAFlnk2
383  chr1  10009400  10009600  TssAFlnk2
384  chr1  10009600  10009800     QuiesG
385  chr1  10009800  10010000     QuiesG
386  chr1  10010000  10010200     QuiesG
>>> filtered.groupby("name", sort=False)
<pandas.core.groupby.generic.DataFrameGroupBy object at 0x7f9da1fe4580>
>>> for name, group in groups:
...   print(group["start"].min(), "\t", group["end"].max(), "\t", group["name"].unique())
...
10007800 	 10008600 	 ['Quies']
10008600 	 10008800 	 ['EnhPois1']
10008800 	 10009600 	 ['TssAFlnk2']
10009600 	 10010200 	 ['QuiesG']
```

Reply:
make `name` a parameter to the function so its more generalized